xbmc/input/InputCodingTableKorean.cpp

```diff
@@ -0,0 +1,371 @@
+
+void CInputCodingTableKorean::SetTextPrev(const std::string& strTextPrev)
+{
+  m_strTextPrev = strTextPrev;
+}
+
+int CInputCodingTableKorean::MergeCode(int choseong, int jungseong, int jongseong)
+{
+  return (unsigned short) 0xAC00 + choseong * 21 * 28 + jungseong * 28 + jongseong + 1;
+}
+
+// Reference
+// https://en.wikipedia.org/wiki/Hangul
+// http://www.theyt.net/wiki/%ED%95%9C%EC%98%81%ED%83%80%EB%B3%80%ED%99%98%EA%B8%B0
+
+std::wstring CInputCodingTableKorean::InputToKorean(const std::wstring& input)
+{
+  std::wstring dicEnglish = L"rRseEfaqQtTdwWczxvgkoiOjpuPhynbml";
+  std::wstring dicKorean = L"ㄱㄲㄴㄷㄸㄹㅁㅂㅃㅅㅆㅇㅈㅉㅊㅋㅌㅍㅎㅏㅐㅑㅒㅓㅔㅕㅖㅗㅛㅜㅠㅡㅣ";
+  std::wstring dicChoseong = L"ㄱㄲㄴㄷㄸㄹㅁㅂㅃㅅㅆㅇㅈㅉㅊㅋㅌㅍㅎ";
+  std::wstring dicJungseong = L"ㅏㅐㅑㅒㅓㅔㅕㅖㅗㅘㅙㅚㅛㅜㅝㅞㅟㅠㅡㅢㅣ";
+  std::wstring dicJongseong = L"ㄱㄲㄳㄴㄵㄶㄷㄹㄺㄻㄼㄽㄾㄿㅀㅁㅂㅄㅅㅆㅇㅈㅊㅋㅌㅍㅎ";
+
+  std::wstring korean;
+  
+  if (input.empty())
+    return korean;
+
+  int choseong = -1, jungseong = -1, jongseong = -1;
+
+  for (unsigned int i = 0; i < input.size(); i++)
+  {
+    wchar_t ch = input.at(i);
+    int key = dicKorean.find(ch);
+
+    // H/W Keyboard input with English will be changed to Korean 
+	// because H/W input in Korean is not supported.
+    if (key == -1)    
+      key = dicEnglish.find(ch);
+
+    if (key == -1) // If not Korean and English
+    {
+      // If there is remained Korean, merge code into character
+      if (choseong != -1) // There is choseong
+      {
+        if (jungseong != -1) // choseong+jungseong+(jongseong)
+          korean += MergeCode(choseong, jungseong, jongseong);
+        else // Only choseong
+          korean += dicChoseong.at(choseong);
+      }
+      else
+      {
+        if (jungseong != -1) // Jungseong
+          korean += dicJungseong.at(jungseong);
+        
+        if (jongseong != -1) // Jongseong
+          korean += dicJongseong.at(jongseong);
+      }
+      choseong = -1;
+      jungseong = -1;
+      jongseong = -1;
+      korean += ch;
+    }
+    else if (key < 19) // If key is consonant, key could be choseong or jungseong
+    {
+      if (jungseong != -1)
+      {
+        if (choseong == -1) // Jungseong without choseong cannot have jongseong. 
+		                    // So inputted key is jungseong character, new character is begun. 
+        {
+          korean += dicJungseong.at(jungseong);
+          jungseong = -1;
+          choseong = key;
+        }
+        else // Jungseong with choseong can have jongseong.
+        {
+          if (jongseong == -1) // Chongseong can have two consonant. So this is first consonant of chongseong.
+          {
+            jongseong = dicJongseong.find(dicKorean.at(key));
+            if (jongseong == -1) // This consonant cannot be jongseong. ex) ㄸ, ㅃ, ㅉ
+			{
+              korean += MergeCode(choseong, jungseong, jongseong);
+              choseong = dicChoseong.find(dicKorean.at(key));
+              jungseong = -1;
+            }
+          }
+          else if (jongseong == 0 && key == 9) 			// ㄳ

```

Comment:
Cosmetics, the comments from here downwards are using tabs instead of spaces and not aligned.